### ISSUE TITLE
PuTTY Pageant

### ISSUE THREAD

**Issue by lars18th** [Microsoft/vscode-remote-release#61]
Hi,

We need to support PuTTY pageant too instead of only plain certificate files.
It will be added soon?

Regards.

**Comment by lars18th** [comment]
Hi,

Until native support will be implement, here is some tool as a workaround (test & working):
https://github.com/benpye/wsl-ssh-pageant

Regards.

**Comment by powerpbx** [comment]
+1 for this feature.  We use pageant for all our SSH private key work.  We tried that workaround tool above on Win10 build 1903 but it didn't seem to work.  Maybe we did something wrong.  Would be nice if it just worked like it does with PuTTy and WinSCP.

**Comment by lars18th** [comment]
> +1 for this feature. We tried that workaround tool above on Win10 build 1903 but it didn't seem to work. Maybe we did something wrong. Would be nice if it just worked like it does with PuTTy and WinSCP.

Hi @powerpbx ,

I'm sure you did something wrong, as I'm using it for _months_ and it works like a charm!
I recommend to read the Documention of the tool, as it **requires** parameters to work.

That's the content of the .BAT that I use to launch it:
```
@ECHO OFF
SETX SSH_AUTH_SOCK \\.\pipe\ssh-pageant
START /B "" "C:\SSH\wsl-ssh-pageant-amd64-gui.exe" --systray --winssh ssh-pageant
```
If you execute it before any other binary, and after you open the Pageant and load your key, then all OpenSSH sessions will use the keys.

I hope it helps you!

**Comment by randallpittman** [comment]
I wrote a short how-to on getting wsl-ssh-pageant to run on startup, followed by pageant. Now I finally get hassle-free public-key auth on VS Code Remote SSH! 

https://github.com/benpye/wsl-ssh-pageant/issues/16#issuecomment-520962345

**Comment by rcdailey** [comment]
I cannot get this working. I followed the instructions and I still do not see Remote SSH trying to use the agent to authenticate with a private key:

```
Install terminal quit with output: The process tried to write to a nonexistent pipe.
```

I'm not sure why it's giving these errors. I verified that the `SSH_AUTH_SOCK` env var is set as well. Which `ssh.exe` is VS Code using?

To complicate things, I actually prefer `gpg-agent.exe` because of its support for Yubikeys. However, I did test with pure pageant and it didn't work either.

**Comment by lars18th** [comment]
Hi @rcdailey 

> I cannot get this working. I followed the instructions and I still do not see Remote SSH trying to use the agent to authenticate with a private key:
> 
> ```
> Install terminal quit with output: The process tried to write to a nonexistent pipe.
> ```
> 
> I'm not sure why it's giving these errors. I verified that the `SSH_AUTH_SOCK` env var is set as well. Which `ssh.exe` is VS Code using?

Instead of trying with VSCode, first try to logging using the SSH client. Please try this:
- Configure your environment (load the wsl-ssh-pageant tool, run the pageant, read the key, set the SSH_AUTH_SOCK env var, configure the '.ssh/config' file, etc.).
- Using your Windows Commnad (CMD or PS) use the Native SSH Client (Windows 10) or the OpenSSH Client (Windows 7, manual install) connect to the server.

If this works, then the problem is in your VSCode-SSH configuration.
If this doens't work, then the problem is in your environment configuration.

I hope it helps! :wink:

**Comment by comphilip** [comment]
`wsl-ssh-pageant` does not work for me neither. ssh with -v option, it shows `wsl-ssh-pageant` not correct respond signature type:

```
debug1: rekey after 134217728 blocks
debug1: SSH2_MSG_NEWKEYS sent
debug1: expecting SSH2_MSG_NEWKEYS
debug1: SSH2_MSG_NEWKEYS received
debug1: rekey after 134217728 blocks
debug1: SSH2_MSG_EXT_INFO received
debug1: kex_input_ext_info: server-sig-algs=<ssh-ed25519,ssh-rsa,rsa-sha2-256,rsa-sha2-512,ssh-dss,ecdsa-sha2-nistp256,ecdsa-sha2-nistp384,ecdsa-sha2-nistp521>
debug1: SSH2_MSG_SERVICE_ACCEPT received
debug1: Authentications that can continue: publickey
debug1: Next authentication method: publickey
debug1: Offering public key: RSA SHA256:1YCSVQTu+76ieXXnbwKDf97dMkLFJRZb/6h6Shb3Ebs cardno:000607309868
debug1: Server accepts key: pkalg rsa-sha2-512 blen 279
warning: agent returned different signature type ssh-rsa (expected rsa-sha2-512)
debug1: Authentications that can continue: publickey
debug1: Trying private key: C:\\Users\\comph/.ssh/id_rsa
debug1: Trying private key: C:\\Users\\comph/.ssh/id_dsa
debug1: Trying private key: C:\\Users\\comph/.ssh/id_ecdsa
debug1: Trying private key: C:\\Users\\comph/.ssh/id_ed25519
debug1: Trying private key: C:\\Users\\comph/.ssh/id_xmss
debug1: No more authentication methods to try.
```

**Comment by lars18th** [comment]
Hi @comphilip ,

> `wsl-ssh-pageant` does not work for me neither. ssh with -v option, it shows `wsl-ssh-pageant` not correct respond signature type:

OK, so the problem isn't the VSCode SSH extension.

Please, try this:
- Inside the same Windows Shell window (CMD or PS) execute all the commands:
  - Set the environment var of SSH_AUTH_SOCK.
  - Start the `wsl-ssh-pageant` with the corresponding parameters `--systray --winssh ssh-pageant`.
  - Execute the pageant and load the key (I recommend too to check the login with the key with PuTTY).
  - Finally execute the OpenSSH Client.

Futhermore:
- Which version of the OpenSSH Client you use? Try last version.
- Ensure that the "ssh-agent" service of the OpenSSH is _disabled_, as it can interfere.

I hope it helps!

**Comment by comphilip** [comment]
Hi @lars18th, 
OpenSSH Client: OpenSSH_for_Windows_7.7p1, LibreSSL 2.6.5
I tried your suggestion and find a strange situation.

It works with `Ubuntu 18.04.3 LTS with OpenSSH_7.6p1 Ubuntu-4ubuntu0.3, OpenSSL 1.0.2n  7 Dec 2017`. This ubuntu is hosted in VirtualBox.

Another 3 remote machines not work: 
- raspberry pi, Raspbian GNU/Linux 10 (buster), OpenSSH_7.9p1 Raspbian-10, OpenSSL 1.1.1c  28 May 2019
- laptop, Debian GNU/Linux 10 (buster), OpenSSH_7.9p1 Debian-10, OpenSSL 1.1.1c  28 May 2019
- server, Container Linux by CoreOS, OpenSSH_7.9p1, OpenSSL 1.0.2r  26 Feb 2019

After turn on DEBUG3 log level of raspberry pi, and found ssh server reject my public key: `debug3: mm_answer_keyverify: publickey 0xbf5288 signature unverified`

I had not idea why. My public key is stored at a Yubikey via OpenPGP protocol. The key works well with gpg-agent alone with ssh-agent in mobxterm.

I believe this is a compatible issue of `wsl-ssh-pageant`.

**Comment by lars18th** [comment]
> I believe this is a compatible issue of `wsl-ssh-pageant`.

I don't think so!

Which Pageant version are you using? Try last version (0.72).

And if you have troubles with the Yubikey, then first try using regular PuTTY keys with `wsl-ssh-pageant`.

Regards.

**Comment by comphilip** [comment]
@lars18th Fixed after upgrade gpg to latest version (2.2.17)

**Comment by lars18th** [comment]
> @lars18th Fixed after upgrade gpg to latest version (2.2.17)

Great! :smile: 

The use of the `wsl-ssh-pageant` tool with any Pageant server is very robust. :wink:

**Comment by Tremolo4** [comment]
**TL;DR:** If you're using the `ssh` and `ssh-pageant` bundled with Git for Windows instead of Windows-native OpenSSH, you just need to add the environment variable below via Windows settings.

`wsl-ssh-pageant` did not work for me, either.

I'm not using the Windows-native OpenSSH client (Win-SSH) at all, instead I use the one bundled with Git for Windows (GfW-SSH). VS Code will fallback to that one if Win-SSH is not installed: IIRC I removed it via power shell, because I needed the JumpHost support from GfW-SSH.

GfW also includes the [`ssh-pageant`](https://github.com/cuviper/ssh-pageant) tool which makes PuTTY pageant available in Git Bash and GfW-SSH. In essence the same thing as `wsl-ssh-pageant`, only for GfW-SSH and not for WSL or Win-SSH. To use it, add the following to your `.bashrc`:
`eval $(/usr/bin/ssh-pageant -r -a "/tmp/.ssh-pageant-$USERNAME")`

However: VS Code still calls the Git for Windows `ssh` from the `cmd.exe`, not from Git Bash, hence the `SSH_AUTH_SOCK` is missing, because it is only in Git Bashs env from your `.bashrc`. Furthermore, the `/tmp` path is mapped to your Temp folder in AppData. In order to use it from `cmd.exe`, i.e. with VS Code Remote, you have to add the following env variable via Windows settings:

`SSH_AUTH_SOCK="%Temp%\.ssh-pageant-%USERNAME%"`

Remember to start PuTTY pageant and open Git Bash at least once, so that `ssh-pageant` starts as well (no need to keep Git Bash open).

**Comment by lars18th** [comment]
Hi @Tremolo4 ,

> `wsl-ssh-pageant` did not work for me, either.

I can't understand why. The only tool that works for me in _any_ Windows PC (7.0 & 10), without complex configurations is with the `wsl-ssh-pageant` tool + _any_ pageant agent (even with smartcard certificates). So for this reason I recommend it.

**Comment by Tremolo4** [comment]
@lars18th It's probably because I am not using the Windows OpenSSH client, but the Git for Windows one.

**Comment by lars18th** [comment]
> @lars18th It's probably because I am not using the Windows OpenSSH client, but the Git for Windows one.

Then I suggest this: install a recent version of the Windows OpenSSH and remove (it's sufficient to rename the `ssh.exe` file) the GIT version. I'm sure this will fix 99% of the troubles.

I hope it helps.

**Comment by Tremolo4** [comment]
Thanks, but I prefer my solution, because I need `ProxyJump`, which was not working last time I tried Windows OpenSSH (see [here](https://github.com/PowerShell/Win32-OpenSSH/issues/1172)).

I'm sure `wsl-ssh-pageant` would work for me if I was using Windows OpenSSH. Just wanted to show a workaround for people like me who can't use Windows OpenSSH for some reason.

**Comment by lars18th** [comment]
Hi @Tremolo4 ,

I use, not `ProxyJump` but `ProxyCommand` with Windows OpenSSH and it works like a charm. So I feel `ProxyJump` will work too. In fact, _Windows_ OpenSSH _is_ OpenSSH, only compiled by Microsoft.

Regards.

**Comment by Tremolo4** [comment]
Yes, I thought it should work, too. But it didn't when I tried, so I tried something else. ¯\\_(ツ)_/¯

**Comment by niklasholm** [comment]
> 
> 
> Hi @Tremolo4 ,
> 
> I use, not `ProxyJump` but `ProxyCommand` with Windows OpenSSH and it works like a charm. So I feel `ProxyJump` will work too. In fact, _Windows_ OpenSSH _is_ OpenSSH, only compiled by Microsoft.
> 
> Regards.

Except full of bugs that OpenSSH on other platforms doesn't have.

**Comment by sdwru** [comment]
Here is a wild and crazy idea.  How about making VSC work with Pageant.  A super simple solution that just works for just about everything except VSC.

**Comment by lars18th** [comment]
Hi,

> A super simple solution that just works for just about everything except VSC.

It works for VSC, as I'm using it for years.

> This discussion is over a year old now and still no proper solution in sight.

A proper solution exist over a lot of time: https://github.com/benpye/wsl-ssh-pageant

Regards.

**Comment by sdwru** [comment]
> Hi,
> 
> > A super simple solution that just works for just about everything except VSC.
> 
> It works for VSC, as I'm using it for years.
> 
> > This discussion is over a year old now and still no proper solution in sight.
> 
> A proper solution exist over a lot of time: https://github.com/benpye/wsl-ssh-pageant
> 
> Regards.

That is a workaround (that many have trouble getting to work).  A proper solution is native support with no fuss.

**Comment by lars18th** [comment]
Hi @sdwru ,

> That is a workaround (that many have trouble getting to work). A proper solution is native support with no fuss.

That's depend on your point of view. It's `ssh.exe` from GIT a workaround?

Please, think on this: any `pageant.exe` is an tool to use certificates. A lot of implementations of it exists. This is a common tool for a lot of Windows tools that use certificates. However, it's not officially used by Microsoft. However, a simple `proxy` tool exist to connect the _standard_ OpenSSH client in Windows to this _the facto standard_ for Windows (I speak about `pageant`). Then the use of a `proxy` is not a workaround? Is not using VSC a proxy when it call to some `ssh.exe` ?

Please, check the opening of this issue: it's opened by me. And the second post is a _working_ solution, but working. No tricks or complex implementations. A simple tool that connects between two executables: the `ssh.exe` and the `pageant`. Everything is open, free and well tested.

I hope it helps.
Regards.

**Comment by Kadeluxe** [comment]
**For future readers who may try to find how to use VS Code Remote and keys from Pageant:**
1. Create .bat file somewhere with the following content:
```
echo OpenSSH
"C:\YOUR_PATH_HERE\PLINK.EXE" -ssh %*
```
2. Open VS Code settings, type `remote ssh path` in search and find **Remote.SSH: Path** settings
3. Past here path to your .bat file
4. Now VS Code Remote will use Pageant correctly.

**Comment by sdwru** [comment]
Thanks for the info 

> **For future readers who may try to find how to use VS Code Remote and keys from Pageant:**
> 
> 1. Create .bat file somewhere with the following content:
> 
> ```
> echo OpenSSH
> "C:\YOUR_PATH_HERE\PLINK.EXE" -ssh %*
> ```
> 
> 1. Open VS Code settings, type `remote ssh path` in search and find **Remote.SSH: Path** settings
> 2. Past here path to your .bat file
> 3. Now VS Code Remote will use Pageant correctly.

Thanks for the info.  I still do not have this working after months of trying but I will try your suggestion.  I basically just gave up and use user/pass for all my development.  Every time there is a new VSC update I anxiously read the changelog to see if they added native support but still nothing.

It's a really annoying problem because I access everything using Pageant and VSC is the only thing that doesn't work with it.  It also does silly things with user/pass.  I have to enter it twice which makes no sense.  But VSC is so nice to use for development I tolerate it.

**Comment by Kadeluxe** [comment]
I was in the same boat as you, I store my keys in KeePass and use its extension KeeAgent.
This method worked for me and now it works flawlessly.

Still, it's probably better to switch to native Windows SSH and key agent but there is no matching functionality as my current setup so I don't care for now.

**Comment by donestefan** [comment]
> For future readers who may try to find how to use VS Code Remote and keys from Pageant:
> 
>     Create .bat file somewhere with the following content:
> 
> echo OpenSSH
> "C:\YOUR_PATH_HERE\PLINK.EXE" -ssh %*
> 
>     Open VS Code settings, type remote ssh path in search and find Remote.SSH: Path settings
>     Past here path to your .bat file
>     Now VS Code Remote will use Pageant correctly.

When using this workaround with stored ssh sessions, make sure you use the following syntax in the [ssh config file](https://code.visualstudio.com/blogs/2019/10/03/remote-ssh-tips-and-tricks#_ssh-configuration-file):
```
Host username1@host1.example.com
Host username2@host2.example.com
```
instead of the regular syntax:
```
Host MyWebserver
    HostName host1.example.com
    User username1
Host MyDBserver
    HostName host2.example.com
    User username2
```

**Comment by lars18th** [comment]
Hi,

As commented in the second post of this issue using a _very simple_ external tool you can do it without troubles: https://github.com/microsoft/vscode-remote-release/issues/61#issuecomment-489307939 .
That solution is more simple than touching the ssh config file.

Regards.

**Comment by Kadeluxe** [comment]
One bat file is _simpler_.

**Comment by lars18th** [comment]
> One bat file is _simpler_.

But it requires some _configuration_ for every project. The alternative is a simple daemon tool that runs in parallel to the pageant daemon and it's transparent. So you can crate a bat file to launch booth with one click and forget!

Anyway, _any_ solution is useful. :wink:

**Comment by Kadeluxe** [comment]
Honestly for me it works without any configuration...
Moreover, I think ssh config file supports wildcard so you can use
```
Host *
  ForwardAgent yes
```

**Comment by sdwru** [comment]
After I did root@host1.example.com it finally worked for me using the *.bat file.  I have been trying to get this working for months.  What a PiTA.  Please add native pageant support Microsoft!

**Comment by Kadeluxe** [comment]
You need to connect once using PuTTY to save the key since it stores them separately from Windows' native SSH, VS Code or whatever.

**Comment by sdwru** [comment]
> You need to connect once using PuTTY to save the key since it stores them separately from Windows' native SSH, VS Code or whatever.

Yes, and that is a hacky non-elegant solution imo.  I am constantly setting up and tearing down servers so this is not really a great solution for me.

**Comment by Kadeluxe** [comment]
Well, you can do
`echo y | plink -ssh root@REMOTE_IP_HERE "exit"` in bat file before actual plink connection.
I mean, it's still a workaround so I don't know how to do it better.

**Comment by sdwru** [comment]
> Well, you can do
> `echo y | plink -ssh root@REMOTE_IP_HERE "exit"` in bat file before actual plink connection.
> I mean, it's still a workaround so I don't know how to do it better.

Thanks for the info.  I was actually just trying to do that.  :)

**Comment by sdwru** [comment]
So now the other problem I have run into is that I cannot use ports other than `22`.  That's kind of a deal breaker for me because I run all my servers on non-standard SSH ports.

Adding `Port` to the config file doesn't work. It's ignored.  Adding it to the end of `Host` doesn't work because VSC converts that to  `-p XXXX` in the command line and plink.exe expects capital `P`.

**Comment by Kadeluxe** [comment]
I guess you can't get away with simple bat file then.
Proper wrapper needed at this point to parse args and convert them to whatever plink wants.

I mean, of course this should be resolved in VS Code directly but for now I'm out of ideas sadly apart from creating a proper program wrapper or patching Plink.

**Comment by sdwru** [comment]
I guess it's back to user/pass authentication for me once again.  I wasted yet another morning messing around with this.  It would be nice if I could get some actual work done using this some day.  I guess that's still just a distant dream.

**Comment by benliddicott** [comment]
A similar batch file workaround works for me **provided** 

* I have an identical session already configured in PuTTY.
* I have Pageant running and the key loaded
* I have already connected to it once and saved the key

Here is the contents of my batch file

```cmd
@echo off
set SCRIPTDIR=%~dp0
if ARG%1==ARG-V goto :VERSION
plink %1 %2 %3 %4 %5 %6 %7 %8 %9

goto :EOF
:VERSION
echo "OpenSSH wrapper for plink"
:EOF
```

And. if I have a .ssh/config entry like this:

```
Host foo-server1
    HostName 192.35.46.57
    User foouser
    ForwardAgent yes
    ForwardX11 yes
    TCPKeepAlive yes
```
Then I have to set up an identical configuration in PuTTY with the same name, and connect to it once.

That done, VS code can connect.

**Comment by Benargee** [comment]
I just keep getting `the remote host may not meet vscode server prerequisites for glibc and libstdc++` everytime I try either of these bash files.
I'm on Win 7.
I can get it working with password authentication with the regular G4W ssh.exe but would really like to get it working with Pageant/KeePass/KeeAgent for automatic key auth. KeeAgent works with PuTTY without issue.

**Comment by Tremolo4** [comment]
@Benargee If you can get it working with G4W, then maybe the bundled `ssh-pageant` will work for you?

In order to make VS Code use G4W `ssh.exe`, you need to uninstall the Windows-bundled OpenSSH (via "Add or remove programs", "Optional features").

Then, to make it use your PuTTY Pageant keys, you need to set-up `ssh-pageant` in your .bashrc and add an env var to make it visible to VS Code. See [my previous comment](https://github.com/microsoft/vscode-remote-release/issues/61#issuecomment-599071804) for details.

**Comment by sdwru** [comment]
This is still a far cry from having native support in VSC.  I've managed to get this working on a couple computers but it was a struggle each time.  I keep checking the changelogs everytime there is a VSC update hoping one day they put it in there.

**Comment by Tremolo4** [comment]
Yeah it's a hassle to set up. But it's also a complicated problem for VS Code to solve completely, since it depends on which `ssh.exe` is used for example. Hard to make sure it doesn't break in some corner case.

**Comment by CoryXie** [comment]
Below 2 steps worked for me:

1) Created a VSC.bat with content and set the **Remote.SSH: Path** settings, both exactly as @Kadeluxe originally wrote.

```
echo OpenSSH
"C:\Program Files\PuTTY\plink.exe" -ssh %*
```

2) Update the ssh config file with below format:

```
Host *
  ForwardAgent yes

Host username@remote.host
```

**Comment by I-am-not-a-number** [comment]
> Host *
>   ForwardAgent yes
> 
> Host username@remote.host

I can't get this to work, ssh is used instead ...

```
[15:32:59.263] Checking ssh with ""C:\test\VSC_SSH.bat" -V"
[15:32:59.265] Got error from ssh: spawn "C:\test\VSC_SSH.bat" ENOENT
[15:32:59.265] The specified path "C:\test\VSC_SSH.bat" is not a valid SSH binary
[15:32:59.265] Checking ssh with "ssh -V"
[15:32:59.281] > OpenSSH_for_Windows_7.7p1, LibreSSL 2.6.5
```

I even added a "-V" switch to the batch file, duplicating the "ssh -V" reply.

Cheers and Thanks :-)

**Comment by benliddicott** [comment]
@I-am-not-a-number check out the batch file in my comment https://github.com/microsoft/vscode-remote-release/issues/61#issuecomment-645950231

**Comment by sergey-somkov** [comment]
This note is important: https://github.com/microsoft/vscode-remote-release/issues/61#issuecomment-645950231
> Then I have to set up an identical configuration in PuTTY with the same name, and connect to it once.

`plink` treats host parameter as a PuTTY saved session name:
```
"C:\Program Files\PuTTY\plink.exe" --help
Plink: command-line connection utility
Release 0.70
Usage: plink [options] [user@]host [command]
       ("host" can also be a PuTTY saved session name)
```
So you just create and save PuTTY session, run `pageant` and use saved session from VSC.

**Comment by bersbersbers** [comment]
Related: https://github.com/billziss-gh/sshfs-win/issues/267#issuecomment-780765181 (and below - still WIP), in particular, https://www.chiark.greenend.org.uk/~sgtatham/putty/wishlist/pageant-named-pipe.html

**Comment by ponchofiesta** [comment]
OpenSSH agent for Windows does not support Smartcards. With support for Pageant we could get Smartcard support too (PuTTY-CAC). +1 for this.

**Comment by SommerEngineering** [comment]
In our case, the scripts / batch files mentioned here did not work. Also the `wsl-ssh-pageant` variant did not work. However, a small Go program helped us: https://github.com/benok/ssh2plink. With it everything worked without any issues. The solution in Go is so simple that Microsoft could deliver this small tool together with the extension.

**Comment by alv000h** [comment]
as stated before and as per clarification.

To keep the plink working, first of all you have to configure a session in putty named exactly like target server you want to connect (use your FQDN or IP), this must login at the target server using preconfigured user and pageant loaded key or predefined ssh key

my tests with .ssh/config configurations seems not work (maybe plink not use this file...), instead you can configure things in putty and store it as putty sessions...

after that, you can place this script over there and configure it in Remote.SSH: Path 

```bat
@echo off
REM TARGET HOST MUST BE a Session saved in putty because it not indicates user when launch ssh command, so putty load that session with user preconfigured
set SCRIPTDIR=%~dp0
if ARG%1==ARG-V goto :VERSION
"C:\Program Files\PuTTY\plink.exe" -ssh -agent %*

goto :EOF
:VERSION
echo "OpenSSH wrapper for plink"
:EOF 
```

**Comment by donestefan** [comment]
"remote ssh" should be [working with pageant version 0.77](https://the.earth.li/~sgtatham/putty/0.77/htmldoc/Chapter9.html#pageant-cmdline-openssh)
Support for "remote WSL" already landed in the [pageant development snapshots](https://tartarus.org/~simon/putty-snapshots/htmldoc/Chapter9.html#pageant-cmdline-unix) , which you can find [here](https://www.chiark.greenend.org.uk/~sgtatham/putty/snapshot.html) (WSL1 only)

**Comment by corford** [comment]
Can confirm @donestefan's comment. Pageant now has in-built support for exposing a pipe to the Windows openssh client. No need for `plink`, `wsl-ssh-pageant` or BAT files. Just follow the simple instructions here: https://the.earth.li/~sgtatham/putty/0.77/htmldoc/Chapter9.html#pageant-cmdline-openssh

**Comment by guldil** [comment]
@corford @DonEstefan i follow indication but my Windows 11 ssh doe's reconize pageAnt agent...

file is correctly created in my c:\Users\<username>\.ssh\ folder and i include it in my c:\Users\<username>\.ssh\config and i have an error : Authenticator provider $SSH_SK_PROVIDER did not resolve;

**Comment by goldfita** [comment]
I was able to get this working using using the .bat wrapper, but I can't figure out a way to do it without putting my password in the .bat file. I'm stuck with putty v0.76 which uses the '-pw' flag.  The remote connection seems to recognize the password prompt, but it doesn't give me a chance to enter a password. If I set `remote.SSH.showLoginTerminal=false`, it opens a dialog at the top which says "Enter password for ...", but it has already sent a password ("Sent password" in the output window)! And it opens another dialog that says "Could not establish connection to ...".

**Comment by binarweb** [comment]
@benliddicott Could you explain how you did it? You said it's been working for months, but I can't get it to work.
The .bat version works for standard port, but doesn't on not standard port.
I'm using the latest VSCode 1.79.2.

**Comment by benliddicott** [comment]
> @benliddicott Could you explain how you did it? You said it's been working for months, but I can't get it to work. The .bat version works for standard port, but doesn't on not standard port. I'm using the latest VSCode 1.79.2.

**Comment by benliddicott** [comment]
> @benliddicott Could you explain how you did it? You said it's been working for months, but I can't get it to work. The .bat version works for standard port, but doesn't on not standard port. I'm using the latest VSCode 1.79.2.

For me, this has been working for a few years now, and I no longer use this solution.

So I can't comment on why it is not working for you.

**Comment by amr-tewfik** [comment]
> Hi,
> 
> We need to support PuTTY pageant too instead of only plain certificate files. It will be added soon?
> 
> Regards.

And by plain you mean putty certificates? ssh-agent works flawlessly.

**Comment by binarweb** [comment]
@amr-tewfik how did you made ssh-agent to work flawlessly? I have the ssh-agent running with the certificates and the latest VSCode still asks me for the certificate password every time.

**Comment by benliddicott** [comment]
@binarweb I find that I no longer need a workaround, the standard Windows SSH now works perfectly.

If you have existing PuTTY keys, you can export them from PuTTY, and use them in Windows SSH in the normal way?

**Comment by binarweb** [comment]
@benliddicott What is "Windows SSH"? I've search for it, no relevant results came up.
Is it a built-in tool in Windows? Does it have pagent? If it doesn't, how do you add your existing keys?

**Comment by amr-tewfik** [comment]
> @amr-tewfik how did you made ssh-agent to work flawlessly? I have the ssh-agent running with the certificates and the latest VSCode still asks me for the certificate password every time.

I asume you already added your key with ssh-agent.
Then you need to run it as a service.
In PowerShell type:
```powershell
Start-Service ssh-agent
Set-Service ssh-agent -StartupType Automatic
```

**Comment by binarweb** [comment]
> > @benliddicott What is "Windows SSH"? I've search for it, no relevant results came up. Is it a built-in tool in Windows? Does it have pagent? If it doesn't, how do you add your existing keys? Thanks!
> 
> "Windows SSH" relates to the windows version of the [openssh](https://learn.microsoft.com/en-us/windows-server/administration/openssh/openssh_overview) command line tools collection, which are included in modern windows versions (see C:\Windows\System32\OpenSSH\*.exe). This is what microsoft tools like vscode use for ssh connections. Connections that work with openssh should automatically work with vscode, too. The `ssh-agent.exe` and `ssh-add.exe` commands can replace pageant (you'll find several guides online). If you don't like these replacements and want to keep using pageant, just follow [these instructions](https://the.earth.li/~sgtatham/putty/0.77/htmldoc/Chapter9.html#pageant-cmdline-openssh). This allows windows ssh (and thus vscode) to use keys from pageant. Make sure to use a current version of pageant and to remove all older workarounds you might have hacked into vscode from the comments above (.bat files, etc...).

It works perfectly. I used https://the.earth.li/~sgtatham/putty/0.77/htmldoc/Chapter9.html#pageant-cmdline-openssh
Thank you!

**Comment by jtmoon79** [comment]
I appreciate the workarounds provided here.  In my case, they are too much work for one of the most basic SSH features.

SSH Key querying should be a checkbox item in the _Remote-SSH_ Extension settings, i.e. "_[ ] Query the running Pageant instance for SSH credentials_" (or something like that)

I hope the VS Code Team would have the premier Windows IDE support the well-established key manager, Pageant, "out of the box".  Running these workarounds is cumbersome, especially when you don't have Administrator access on the Windows PC, and/or you're managing multiple Windows PCs that have "floating developers".

**Comment by suningzhao1126** [comment]
> well-established SSH key manager for Windows and should be su

Hi, I have a problem while connecting Linux remote server with keys generated from PuTTY. My key is DSA 1024 format. I can connect to the server using PuTTY perfectly, but I struggle to connect with VS code. 

If I convert the key using PuTTYgen, it does not work. It keeps asking me to enter the password, but I didn't have that. If I tried to generate a "pageant.conf" and include that, it still ask me to enter the password but I have my key in Pageant already.

Felt upset and would like to know how to really solve it. I am a beginner to Linux server. Thanks!

**Comment by powerpbx** [comment]
> > well-established SSH key manager for Windows and should be su
> 
> Hi, I have a problem while connecting Linux remote server with keys generated from PuTTY. My key is DSA 1024 format. I can connect to the server using PuTTY perfectly, but I struggle to connect with VS code.
> 
> If I convert the key using PuTTYgen, it does not work. It keeps asking me to enter the password, but I didn't have that. If I tried to generate a "pageant.conf" and include that, it still ask me to enter the password but I have my key in Pageant already.
> 
> Felt upset and would like to know how to really solve it. I am a beginner to Linux server. Thanks!

Are you selecting `Conversion > Import Key` in PuttyGen?

If you open the DSA key in a text editor, what does it say underneath the `-----BEGIN DSA PRIVATE KEY-----` line?

If it says `Proc-Type: 4,ENCRYPTED` then your DSA key uses a passphrase and that is why PuttGen is asking for a password when you try import it.

**Comment by powerpbx** [comment]
Btw, Windows OpenSSH can talk directly to recent versions of Pageant, so I don't think there is any need to use these 3rd party tools anymore.

https://the.earth.li/%7Esgtatham/putty/0.78/htmldoc/Chapter9.html#pageant-cmdline-openssh

**Comment by suningzhao1126** [comment]
> Btw, Windows OpenSSH can talk directly to recent versions of Pageant, so I don't think you need to use these 3rd party tools anymore. I can confirm the instructions at the following link work with VSC and Pageant v0.81
> 
> https://the.earth.li/%7Esgtatham/putty/0.78/htmldoc/Chapter9.html#pageant-cmdline-openssh

If I did this, do I still need to translate my private key? I have tried this but it seems not work...the program is still trying to open private keys.

**Comment by powerpbx** [comment]
> > Btw, Windows OpenSSH can talk directly to recent versions of Pageant, so I don't think you need to use these 3rd party tools anymore. I can confirm the instructions at the following link work with VSC and Pageant v0.81
> > https://the.earth.li/%7Esgtatham/putty/0.78/htmldoc/Chapter9.html#pageant-cmdline-openssh
> 
> If I did this, do I still need to translate my private key? I have tried this but it seems not work...the program is still trying to open private keys.

This works with PuTTY keys loaded in to Pageant.  No OpenSSH keys required.
